### ISSUE TITLE
[QUERY] Refreshing the JWT token obtained from Azure and updating it in the SecurityContextHolder

### ISSUE THREAD

**Issue by martfol** [Azure/azure-sdk-for-java#34691]
**Query/Question**
Hi guys, I'd like to ask for help regarding the following situation:

We have two services (spring boot apps) - A and B.

- both are registered within our Azure AD and have different client IDs
- both use _AadResourceServerWebSecurityConfigurerAdapter_ to secure the rest interfaces (all "/api/**" paths) and _AadWebSecurityConfigurerAdapter_ to secure all other paths (i.e. the UI)

Then the process flow looks as follows:

1. User logs into the UI of application A (_AadWebSecurityConfigurerAdapter_ takes care of the Azure login process)
2. User uses then the UI of application A and at some point, the backend of A needs to call the rest API of B, so...
3. ... the client in A reads the JWT value of the current user from the principal object (Oidc User) stored in the SecurityContextHolder
4. Then the client in A adds an authorization header with value formatted as "Bearer <token_value>" to the request and sends it to /api of application B
5. B receives the request, validates and processes it - until now the usual scenario works fine, but...

... when we leave the browser session open and continue working after >65 minutes (=token expiration period configured in Azure), and again the client A wants to call backend B, apparently the client in backend A still uses the same JWT token (i.e. it's still in the SecurityContextHolder), which is however past its expiration date and therefore the backend B refuses the token with 401 saying "Failed to authenticate since the JWT was invalid / Jwt expired at xxxxxxxx".

So my question is - what would be the proper way/best practice to do in this situation? Is it possible to achieve that the backend of A automatically recognizes that the current token has alrady expired and performs the refresh and updates the principal in the SecurityContextHolder?

Or do we have to implement it manually, i.e. client A checks the expiration date of the token after reading it from the SecurityContextHolder and when already expired, then refreshes it and updates it somehow in the SecurityContextHolder? If it's this case, then could you please provide link or example code on how to programmatically perform the token refresh with Azure AD?

Thanks a lot in advance for your help and support!

***Why is this not a Bug or a feature Request?***
This is neither a bug, nor a feature request... Just a question :)

**Setup (please complete the following information if applicable):**
 - Library/Libraries: spring-boot 2.7.8, spring-cloud-azure 4.6.0
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by martfol** [comment]
Hi @joshfree, thanks for your reply. I'll be waiting :)

**Comment by martfol** [comment]
Hi guys, any thoughts on this, please?

**Comment by Netyyyy** [comment]
Hi @martfol 
Thank you for reporting this issue. We appreciate your input and will review this matter as soon as possible.
Please feel free to provide any additional information or context that you think may be helpful. We'll keep you updated on the progress of our review.
Thank you for your contribution to improving our project.

**Comment by martfol** [comment]
Hi, @Netyyyy - yes, please, your reply/help will be highly appreciated, as we are still blocked on this!
I managed to google some old tips for spring boot mentioning setting the _reuseRefreshTokens_ attribute to false, but I don't know where to set it or which class/method to override. I was not able to find any official documentation regarding this attribute or setting. Please provide us with some tips or example code on how to solve this problem in spring-boot 2.7.8/spring-cloud-azure 4.6.0. Thanks a lot in advance! (FYI @joshfree)

**Comment by Netyyyy** [comment]
> Hi, @Netyyyy - yes, please, your reply/help will be highly appreciated, as we are still blocked on this! I managed to google some old tips for spring boot mentioning setting the _reuseRefreshTokens_ attribute to false, but I don't know where to set it or which class/method to override. I was not able to find any official documentation regarding this attribute or setting. Please provide us with some tips or example code on how to solve this problem in spring-boot 2.7.8/spring-cloud-azure 4.6.0. Thanks a lot in advance! (FYI @joshfree)

@moarychan please help take a look

**Comment by martfol** [comment]
Hi guys, let's put this topic on hold, it seems that we've been obtaining the access token in a wrong way. We've been reading it from the SecurityContextHolder's principal object, but instead now I see we're supposed to obtain it using the OAuth2AuthorizedClientManager. Maybe that's the reason, will post an update here soon...
